tests/recordCallsToSetTimeout.ts

```diff
@@ -0,0 +1,193 @@
+import EventEmitter from 'events';
+import EMPTY_FUNCTION from './emptyFunction';
+
+type OriginalSetTimeout = typeof global.setTimeout;
+
+type SetTimeoutCallback = () => any;
+
+interface SetTimeoutCall {
+  callback: SetTimeoutCallback;
+  duration: number;
+  timeout: NodeJS.Timeout;
+}
+
+type InterceptingCallback = (
+  callback: SetTimeoutCallback,
+  stopPassingThroughCalls: () => void,
+) => SetTimeoutCallback;
+
+/**
+ * A class that provides a mock implementation for `setTimeout` which records
+ * the callback given so that it can be replayed later.
+ */
+class SetTimeoutRecorder {
+  private originalSetTimeout: OriginalSetTimeout;
+
+  private interceptCallback: InterceptingCallback;
+
+  public calls: SetTimeoutCall[];
+
+  private events: EventEmitter;
+
+  private numAutomaticCallsRemaining: number;
+
+  constructor({
+    originalSetTimeout,
+    numAutomaticCalls = 0,
+    interceptCallback = (callback) => callback,
+  }: {
+    originalSetTimeout: OriginalSetTimeout;
+    numAutomaticCalls?: number;
+    interceptCallback?: InterceptingCallback;
+  }) {
+    this.originalSetTimeout = originalSetTimeout;
+    this.interceptCallback = interceptCallback;
+
+    this.calls = [];
+    this.events = new EventEmitter();
+    this.numAutomaticCallsRemaining = numAutomaticCalls;
+  }
+
+  /**
+   * Looks for the first `setTimeout` call in the call stack that matches the
+   * given duration and calls it, removing it from the call stack, or waits
+   * until such a call appears.
+   *
+   * @param duration - The expected duration of a `setTimeout` call.
+   * @returns A promise that resolves when a `setTimeout` call matching the
+   * given duration is called.
+   */
+  nextMatchingDuration(duration: number): Promise<void> {
+    return new Promise<void>((resolve) => {
+      const index = this.calls.findIndex((call) => call.duration === duration);
+
+      if (index === -1) {
+        const listener = (call: SetTimeoutCall, callIndex: number) => {
+          if (call.duration === duration) {
+            call.callback();
+            this.calls.splice(callIndex, 1);
+            this.events.off('setTimeoutAdded', listener);
+            resolve();
+          }
+        };
+        this.events.on('setTimeoutAdded', listener);
+      } else {
+        this.calls[index].callback();
+        this.calls.splice(index, 1);
```

Comment:
Should this come before the previous line? I was thinking, if this callback triggers something that also calls `nextMatchingDuration`, this might end up doing the wrong thing (i.e. it might be at a different index than before). If that happened it would be very tricky to debug.

Reply:
The same goes for line 68 -  we should remove the call before calling the callback.

Reply:
Good catch! Changed in [c127002](https://github.com/MetaMask/eth-block-tracker/pull/103/commits/c127002db47ee5efddf7d42ee9cd05b33f045d73).